Add support for IRC invite

Vilkku opened the pull request:
The IRC channel I want to mirror with slack is invite-only. This change makes slack-irc respond to invite commands, automatically joining any channel it is invited to.

ekmartin commented:
Good idea! Maybe it should check that the channel exists in the channel mapping, available at `this.channelMapping`, before joining? Kind of bad if you can make the bot join an arbitrary amount of channels by just inviting it.

This [test](https://github.com/ekmartin/slack-irc/blob/master/test%2Fbot.test.js#L33) needs to be updated too.

Vilkku commented:
That's probably a good idea. And forgot about the tests, my bad.

I'm currently adding features that allow me to change usermodes and log in on some various IRC networks - I think the best way to go about it is to add another field to the server settings where you can specify IRC commands that are set on connect.

ekmartin commented:
Yep, that could work. I thought about adding SASL support, but thought it might just be overkill as not all networks support it anyway.

Vilkku commented:
I fixed the test and added the check you suggested.

I also added support for automatically sending commands on connecting, they are added to the server config like this:

    "autosendcmd": [
        ["MODE", "slack", "+x"],
        ["AUTH", "slack", "password"]
    ]

It uses the send command in node-irc, so you can do pretty much everything with it.

Sample configs are not updated. I'm not even sure you like my implementation of the feature, and I should have left it out of this pull request, but I didn't think that far ahead before making the changes :)

ekmartin commented on lib/bot.js:
```diff
@@ -33,6 +33,8 @@ function Bot(options) {
     channels: this.channels
   });
 
+  this.client.autosendcmd = options.autosendcmd;
```
No need to set this on the client, just do `this.autoSendCommands = options.autoSendCommands;` instead (`autosendcmd` -> `autoSendCommands` too), and move it up to the other option-specific assignments (line 25 etc.).

Vilkku replied:
Done. Am I mistaken in thinking that I needed to call bind() on addListener()?

ekmartin commented on lib/bot.js:
```diff
@@ -40,6 +42,9 @@ function Bot(options) {
 Bot.prototype.attachListeners = function() {
   this.client.addListener('registered', function(message) {
     logger.debug('Registered event: ', message);
+    for (var i = 0, len = this.autosendcmd.length; i < len; i++) {
```
Please use a `.forEach`-loop instead :). If you use `.bind` on the listener callback you can use `this.autoSendCommands` even with the changes I mentioned in the comment above.

Vilkku replied:
Done. I did a bind() here as well, is that needed?

ekmartin replied:
Yep, but you can pass `this` as the second argument to `forEach` instead:
```js
    this.autoSendCommands.forEach(function(command) {
      this.client.send.apply(this.client, command);
    }, this);
```
Also you don't need to specify all three `forEach` arguments, since you're only using the first :)

ekmartin commented on lib/bot.js:
```diff
@@ -50,6 +55,16 @@ Bot.prototype.attachListeners = function() {
+
+  this.client.addListener('invite', function(channel, from, message) {
+    logger.debug('Received invite', channel, from, message);
+    if (this.opt.channels.indexOf(channel) > -1) {
```
Would probably be better to do `if (this.channelMapping[channel]) {` here instead, note that you need to do `.bind(this)` on the listener callback then so that `this` is the Bot-instance and not the IRC client itself.

Vilkku replied:
Is that possible to use in this case - channel here is just he name of the IRC channel, while channelMapping seems to take a mapping from the config file as the argument? I left it unchanged for now.

ekmartin replied:
Yep, forgot that the `channelMapping` keys where Slack-channels and not IRC ones.
You can use `this.channels`, from [here](https://github.com/ekmartin/slack-irc/blob/8003e0b8841b367ed125cd8dd35ad6f7af50da19/lib/bot.js#L27), which will be an array of all the IRC channels from the channel mapping, instead of `this.opt.channels`.

ekmartin commented:
I like it, added some feedback :)

Vilkku commented:
Did improvements based on your latest feedback. Thanks for pointing some things out, never really had to do anything this complicated (if you can call it that) in JS.

ekmartin commented on lib/bot.js:
```diff
@@ -40,7 +41,10 @@ function Bot(options) {
   this.client.addListener('registered', function(message) {
     logger.debug('Registered event: ', message);
-  });
+    this.autoSendCommands.forEach(function(element) {
```
This'll crash the bot if you don't provide an `autoSendCommands`-array, can fix it by changing line 28 to
```js
this.autoSendCommands = options.autoSendCommands || [];
```

so it defaults to an empty array if it doesn't exist.

Vilkku replied:
Fixed. I imagined the tests would have failed since they didn't have autoSendCommands specified either, my bad.

ekmartin commented:
Thanks! I'll update the readme and release a new version.

ekmartin commented:
https://github.com/ekmartin/slack-irc/commit/260a5d56a844e735132ba05839060c97d3eac42d
